### PR TITLE
fix(skip-link): work with absolute and relative paths

### PR DESCRIPTION
**Pull request by straker**
Closes issue: #2791

### CODE REVIEW

**Comment by straker** [comment]
I approve for security but clicking the button in the check isn't working for me

File: lib/commons/dom/is-skip-link.js [hunk 1]
```diff
@@ -13,22 +16,41 @@ const isInternalLinkRegex = /^\/?#[^/!]/;
...
   } else {
-    // define a skip link as any anchor element whose href starts with `#...`
-    // and which precedes the first anchor element whose href doesn't start
-    // with  `#...` (that is, a link to a page)
+    // define a skip link as any anchor element whose resolved href
+    // resolves to the current page and uses a fragment identifier (#)
+    // and which precedes the first anchor element whose resolved href
+    // does not resolve to the current page or that doesn't use a
+    // fragment identifier
+    const currentPage = window.location.origin + window.location.pathname;
```
Comment (WilcoFiers) [inline lib/commons/dom/is-skip-link.js:28]: I thought we couldn't count on window.location existing. Won't this throw in JSDOM?

File: lib/commons/dom/is-skip-link.js [hunk 2]
```diff
@@ -13,22 +16,41 @@ const isInternalLinkRegex = /^\/?#[^/!]/;
...
     firstPageLink = querySelectorAll(
-      // TODO: es-module-_tree
       axe._tree,
-      'a:not([href^="#"]):not([href^="/#"]):not([href^="javascript"])'
-    )[0];
+      'a[href]:not([href^="javascript"])'
```
Comment (WilcoFiers) [inline lib/commons/dom/is-skip-link.js:31]: Should that be this: 

```suggestion
      'a[href]:not([href^="javascript:"])'
```

It's unlikely, but no reason to think someone can't just have `<a href="javascript/home.php">` or whatever.

File: lib/commons/dom/is-skip-link.js [hunk 3]
```diff
@@ -13,22 +16,41 @@ const isInternalLinkRegex = /^\/?#[^/!]/;
...
     firstPageLink = cache.get('firstPageLink');
   } else {
-    // define a skip link as any anchor element whose href starts with `#...`
-    // and which precedes the first anchor element whose href doesn't start
-    // with  `#...` (that is, a link to a page)
+    // define a skip link as any anchor element whose resolved href
+    // resolves to the current page and uses a fragment identifier (#)
+    // and which precedes the first anchor element whose resolved href
+    // does not resolve to the current page or that doesn't use a
+    // fragment identifier
```
Comment (WilcoFiers) [inline lib/commons/dom/is-skip-link.js:27]: If you think an explanation like this is needed, create a function and put it in a docblock. Random blocks of comments like these are disruptive.

**Review by WilcoFiers — DISMISSED** [review]
Couple small things.

File: lib/commons/dom/is-skip-link.js (hunk 1 above)
Comment (straker) [reply lib/commons/dom/is-skip-link.js:28]: `window.origin` doesn't exist in IE11, so we can't use that, but `window.location` exists in IE11 and JSDOM. It's just that JSDOM doesn't have any data for the location properties so we can never determine if `path/of/url#id` is a skip link or not.

**Review by Boone6969 — APPROVED** [review]

File: lib/commons/dom/is-skip-link.js (hunk 1 above)
Comment (WilcoFiers) [reply lib/commons/dom/is-skip-link.js:28]: OK, so I think what we do is just feature test. If we can do this check, we do it. If not (like in JSDOM) we fall back to the current behavior.

File: test/node/jsdom.js [hunk 4]
```diff
@@ -58,4 +59,18 @@ describe('jsdom axe-core', function() {
...
   });
+
+  describe('isCurrentPageLink', function() {
+    it('should return null when url is not set', function() {
+      var dom = new jsdom.JSDOM(domStr);
+      var anchor = dom.window.document.getElementById('#skip');
+      assert.strictEqual(axe.commons.dom.isCurrentPageLink(anchor), null);
+    });
+
+    it('should return true when url is set', function() {
```
Comment (WilcoFiers) [inline test/node/jsdom.js:70]: This doesn't check for "true"

File: lib/commons/dom/is-current-page-link.js [hunk 5]
```diff
@@ -12,8 +12,9 @@ const angularRouterLinkRegex = /^#[!/]/;
  * @return {Boolean|null}
  */
 export default function isCurrentPageLink(anchor) {
-  // jsdom can have window.location.origin set to null
-  if (!window?.location.origin) {
+  // jsdom can have window.location.origin set to "null" (the string)
+  // if the url option is not set when parsing the dom string
+  if (window.location.origin === 'null') {
```
Comment (WilcoFiers) [inline lib/commons/dom/is-current-page-link.js:17]: This doesn't look very robust. How about checking for a protocol?

```suggestion
  if (typeof window.location?.origin !== 'string' || window.location.origin.indexOf('://') === -1) {
```

**Review by WilcoFiers — CHANGES_REQUESTED** [review]

**Review by WilcoFiers — APPROVED** [review]

**Pull request merged**